### ISSUE TITLE
Operator diagonal assembly after https://github.com/CEED/libCEED/pull/889

### ISSUE THREAD

**Issue by sebastiangrimberg** [CEED/libCEED#1195]
https://github.com/CEED/libCEED/pull/889 introduced `CeedElemRestrictionCreateOriented` in order to support H(div) and low-order `p=1` H(curl) elements. I ran into an issue realizing that this breaks `CeedOperatorLinearAssembleDiagonal`, due to the final `CeedElemRestrictionApply` (transpose) call at the end of `CeedSingleOperatorAssembleAddDiagonal_Core`. For an oriented basis, this needs to be an "unsigned" apply (transpose) in order to give the correct diagonal entries. Likewise signs on the element restriction need to be handled elsewhere in `ceed-preconditioning.c` like in full assembly.

This will be even further complicated by a solution like https://github.com/CEED/libCEED/pull/1168 for high-order H(curl) support and I haven't quite figured out the best way to get the operator diagonal in this case. In this case, off-diagonal entries of `B^T D B` are going to contribute to the operator diagonal due to pairs of triangular face dofs for Nedelec spaces of order `p > 1`.

For now, I have two questions:

1. For the oriented element restriction case, should we add an `CeedElemRestrictionApplyUnsigned` method which defaults to just `CeedElemRestrictionApply` but for `CeedElemRestrictionCreateOriented` will perform an apply without the orientation signs? This will touch a lot of code and so I wanted to see if this is a good approach before starting. An alternative I see would be to modify the function signature of `CeedElemRestrictionApply` to include an `unsigned` flag but this might be worse.
2. For high-order H(curl) support, do you have any ideas of a best way to approach the issue of diagonal assembly? It seems as though in general we will need to assemble a matrix product `T^T (B^T D B) T` where `T` is the tri-diagonal part of the element restriction accounting for orientations and coupling of pairs of face dofs. I think because of the block structure of `T` (not only is it tri-diagonal, but actually it is diagonal + block 2x2 for the pairs of face degrees of freedom), this means we can get away with constructing only the tri-diagonal part of `B^T D B`. But I don't know how such an interface might be handled such that it can be abstracted away easily enough for the user who just wants to call `CeedOperatorLinearAssembleDiagonal`. I'd appreciate any suggestions or input here.

Thank you!

**Comment by rezgarshakeri** [comment]
Yeah, I discussed with Jed couple of months ago that `CeedSingleOperatorAssembleAddDiagonal_Core` is not gonna work for H(div) (we need something like what you mentioned `E^T (B^T D B) E` ). I created a simple test at that time for one element [here](https://github.com/CEED/libCEED/tree/rezgar/Hdiv-diagonal-assembly) that breaks because of the sign problem, but I got really busy with mixed-elasticity and forgot to report this issue.
